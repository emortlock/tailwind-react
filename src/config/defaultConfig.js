export default {
  radius: 'rounded',
  spacing: {
    zero: 0,
    sm: 2,
    md: 4,
    lg: 8,
  },
  container: {
    sm: 'xl',
    md: '3xl',
    lg: '4xl',
  },
  text: {
    size: {
      body: ['sm', 'base', 'lg'],
      title: ['lg', 'xl', '2xl', '3xl', '4xl', '5xl'],
    },
    family: {
      body: 'sans',
      subtitle: 'sans',
      title: 'sans',
    },
  },
  baseColors: {
    primaryLight: 'blue-light',
    primary: 'blue',
    primaryDark: 'blue-dark',
    secondaryLight: 'grey',
    secondary: 'grey-dark',
    secondaryDark: 'grey-darker',
    successLight: 'green-light',
    success: 'green',
    successDark: 'green-dark',
    dangerLight: 'red-light',
    danger: 'red',
    dangerDark: 'red-dark',
    warningLight: 'orange-lighter',
    warning: 'orange-light',
    warningDark: 'orange',
    infoLight: 'teal-lighter',
    info: 'teal-light',
    infoDark: 'teal',
  },
  textColors: {
    body: 'grey-darkest',
    link: 'blue-dark',
    linkDark: 'blue-darker',
    emphasis: 'black',
    on: {
      primaryLight: 'white',
      primary: 'white',
      primaryDark: 'white',
      secondaryLight: 'white',
      secondary: 'white',
      secondaryDark: 'white',
      successLight: 'white',
      success: 'white',
      successDark: 'white',
      dangerLight: 'white',
      danger: 'white',
      dangerDark: 'white',
      warningLight: 'black',
      warning: 'black',
      warningDark: 'black',
      infoLight: 'black',
      info: 'black',
      infoDark: 'black',
    },
  },
}
