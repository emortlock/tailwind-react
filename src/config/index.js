export { default as defaultConfig } from './defaultConfig'
export { default as TailwindConfig } from './TailwindConfig'
export { default as withConfig } from './withConfig'
