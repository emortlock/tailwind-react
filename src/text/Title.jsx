import React from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'

import withConfig from '../withConfig'

const Title = ({
  config,
  children,
  className,
  is,
  size,
  subtitle,
  flush,
  h,
  ...rest
}) => {
  const level = h || Math.max(7 - size, 1)
  const Component = is || `h${level}`
  let ariaProps = {}

  if (!subtitle && Component !== 'string' && !/h[1-6]/i.test(Component)) {
    ariaProps = {
      role: 'heading',
      'aria-level': level,
    }
  }
  return (
    <Component
      {...rest}
      {...ariaProps}
      className={classnames(
        `font-${config.text.family[subtitle ? 'subtitle' : 'title']}`,
        'leading-tight',
        `text-${config.text.size.title[size - 1]}`,
        !subtitle && `text-${config.textColors.emphasis} font-bold`,
        subtitle && `text-${config.textColors.body} font-medium`,
        !flush && `mb-${config.spacing.md}`,
        className,
      )}
    >
      {children}
    </Component>
  )
}

Title.propTypes = {
  config: PropTypes.shape({}).isRequired,
  children: PropTypes.node,
  className: PropTypes.string,
  is: PropTypes.oneOfType([PropTypes.string, PropTypes.func, PropTypes.object]),
  size: PropTypes.number,
  subtitle: PropTypes.bool,
  flush: PropTypes.bool,
  h: PropTypes.number,
}

Title.defaultProps = {
  children: undefined,
  className: undefined,
  is: undefined,
  size: 4,
  subtitle: false,
  flush: false,
  h: undefined,
}

export { Title as component }
export default withConfig(Title)
