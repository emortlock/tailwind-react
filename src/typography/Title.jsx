import React from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'

import { withTheme } from '../theme'
import {
  getTailwindClassNames,
  tailwindProps,
  tailwindPropTypes,
} from '../tailwind'
import { filterProps } from '../utils'

const Title = ({
  theme,
  children,
  className,
  is,
  size,
  subtitle,
  flush,
  h,
  ...rest
}) => {
  const level = h || Math.max(7 - size, 1)
  const Component = is || `h${level}`
  const userClassNames = classnames(getTailwindClassNames(rest), className)

  let ariaProps = {}

  if (!subtitle && Component !== 'string' && !/h[1-6]/i.test(Component)) {
    ariaProps = {
      role: 'heading',
      'aria-level': level,
    }
  }
  return (
    <Component
      {...filterProps(rest, tailwindProps)}
      {...ariaProps}
      className={classnames(
        `font-${theme.text.family[subtitle ? 'subtitle' : 'title']}`,
        'leading-tight',
        `text-${theme.text.size.title[size - 1]}`,
        !subtitle && `text-${theme.textColors.emphasis} font-bold`,
        subtitle && `text-${theme.textColors.body} font-medium`,
        !flush && `mb-${theme.spacing.md}`,
        userClassNames,
      )}
    >
      {children}
    </Component>
  )
}

Title.propTypes = {
  theme: PropTypes.shape({}).isRequired,
  children: PropTypes.node,
  className: PropTypes.string,
  is: PropTypes.oneOfType([PropTypes.string, PropTypes.func, PropTypes.object]),
  size: PropTypes.number,
  subtitle: PropTypes.bool,
  flush: PropTypes.bool,
  h: PropTypes.number,
  ...tailwindPropTypes,
}

Title.defaultProps = {
  children: undefined,
  className: undefined,
  is: undefined,
  size: 4,
  subtitle: false,
  flush: false,
  h: undefined,
}

export { Title as component }
export default withTheme(Title)
