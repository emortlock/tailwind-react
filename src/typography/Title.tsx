import type { TextProps } from 'tailwind-react-primitives'
import type { WithTheme } from '../theme'

import React from 'react'

import { RawText as Text } from 'tailwind-react-primitives'
import { withTheme } from '../theme'

export interface TitleProps<E extends HTMLElement = HTMLHeadingElement>
  extends WithTheme,
    Omit<TextProps<E>, 'size'> {
  subtitle?: boolean
  flush?: boolean
  level?: number
  size?: number
}

export const RawTitle = <E extends HTMLElement = HTMLHeadingElement>({
  theme,
  children,
  as = 'h1',
  size = 4,
  subtitle,
  flush,
  level = 1,
  ...rest
}: TitleProps<E>) => {
  const hLevel = level || Math.max(7 - size, 1)
  const element: TitleProps['as'] = as || `h${hLevel}`

  let ariaProps: TextProps<E> = {}

  if (!subtitle && typeof element === 'string' && !/h[1-6]/i.test(element)) {
    ariaProps = {
      'role': 'heading',
      'aria-level': hLevel,
    }
  }

  return (
    <Text<E>
      as={element}
      {...ariaProps}
      leading="tight"
      font={[
        theme.text.family[subtitle ? 'subtitle' : 'title'],
        subtitle ? 'medium' : 'bold',
      ]}
      text={[
        theme.text.size.title[size - 1],
        subtitle ? theme.textColors.body : theme.textColors.emphasas,
      ]}
      m={!flush ? { b: theme.spacing.md } : undefined}
      {...rest}
    >
      {children}
    </Text>
  )
}

export const Title = withTheme<TitleProps>(RawTitle)
