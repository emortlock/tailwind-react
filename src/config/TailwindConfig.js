import React from 'react'
import defaultConfig from './defaultConfig'

export default React.createContext(defaultConfig)
