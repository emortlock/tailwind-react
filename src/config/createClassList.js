import merge from 'lodash.merge'
import defaultConfig from './defaultConfig'

const mapObject = (obj, prefix) =>
  Object.keys(obj).map(key => `${prefix}-${obj[key]}`)

const mapArray = (arr, prefix) => arr.map(value => `${prefix}-${value}`)

const reduceMultiplePrefixes = (obj, prefixes) =>
  Object.keys(obj).reduce(
    (arr, key) =>
      typeof obj[key] === 'object'
        ? [...arr, ...reduceMultiplePrefixes(obj[key], prefixes)]
        : [...arr, ...prefixes.map(prefix => `${prefix}-${obj[key]}`)],
    [],
  )

export default (projectConfig = {}) => {
  const config = merge(defaultConfig, projectConfig)

  return [
    config.radius,
    ...reduceMultiplePrefixes(config.spacing, [
      'mb',
      'mr',
      '-ml',
      '-mr',
      'p-',
      'pl-',
      'px-',
      'py-',
    ]),
    ...mapObject(config.container, 'max-w'),
    ...mapArray(config.text.size.body, 'text'),
    ...mapArray(config.text.size.title, 'text'),
    ...mapObject(config.text.family, 'font'),
    ...reduceMultiplePrefixes(config.baseColors, [
      'background-color',
      'border-color',
      'text',
    ]),
    ...reduceMultiplePrefixes(config.textColors, [
      'background-color',
      'border-color',
      'text',
    ]),
  ].filter((className, index, self) => self.indexOf(className) === index)
}
