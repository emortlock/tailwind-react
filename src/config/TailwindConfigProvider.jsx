import React from 'react'
import PropTypes from 'prop-types'
import merge from 'lodash.merge'

import defaultConfig from './defaultConfig'
import TailwindConfig from './TailwindConfig'

const TailwindConfigProvider = ({ config, children }) => {
  const mergedConfig = merge(defaultConfig, config)

  return (
    <TailwindConfig.Provider value={mergedConfig}>
      {children}
    </TailwindConfig.Provider>
  )
}

TailwindConfigProvider.propTypes = {
  config: PropTypes.shape({}),
  children: PropTypes.node,
}

TailwindConfigProvider.defaultProps = {
  config: {},
  children: undefined,
}

export default TailwindConfigProvider
