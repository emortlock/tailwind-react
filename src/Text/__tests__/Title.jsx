import React from 'react'
import { shallow } from 'enzyme'

import { component as Title } from '../Title'
import { defaultConfig } from '../../config'

const setup = (testProps = {}) => {
  const props = Object.assign({ config: defaultConfig }, testProps)

  const wrapper = shallow(<Title {...props}>Title</Title>)

  return {
    props,
    wrapper,
  }
}

describe('Title', () => {
  it('renders matching snapshot', () => {
    const { wrapper } = setup({ size: 3 })

    expect(wrapper).toMatchSnapshot()
  })
})
