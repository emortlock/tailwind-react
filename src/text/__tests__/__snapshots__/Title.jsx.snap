// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Title renders matching snapshot 1`] = `
<h4
  className="font-sans leading-tight text-2xl text-black font-bold mb-4"
>
  Title
</h4>
`;
